Issue #140: Prototype session storage mechanism

csadilek opened the pull request:
This adds a prototype of a session storage mechanism that's capable of storing both browser and engine session state (and the combination) using a configurable session storage. It supports periodic saves at a configurable interval, and restores the persisted state on startup. A default storage mechanism is provided, which stores the session state in a JSON file as well as in memory.

This includes a small refactoring/simplification of our existing session concepts (hence the number of affected files). When implementing this, it didn't feel right to introduce yet another session concept that uses `SessionMapping`, `SessionManager`, and `SessionProvider`. Since we were not really happy with `SessionMapping` anyway, I've simplified these concepts so there's now a `SessionProvider` as part of feature-session which covers the `SessionMapping` functionality and which contains a `SessionManager`. Looking at our updated [sample app](https://github.com/csadilek/android-components/blob/issue-140/samples/browser/src/main/java/org/mozilla/samples/browser/Components.kt) and [use cases](https://github.com/csadilek/android-components/blob/issue-140/components/feature/session/src/main/java/mozilla/components/feature/session/SessionUseCases.kt), this does simplify the API quite a bit, as we no longer have to pass around a `SessionManager` and `SessionMapping`, just a `SessionProvider`, which also takes care of storage concerns.

Another decision here is that the storage makes use of our `SessionProxy` directly. I think this makes sense as `SessionProxy` is a combination of `Session` and `EngineSession` and represents their combined state.

Currently this only persists and restores the URLs of all actives sessions and the selected session id, but this would be good enough to land indepentently and we can address open questions in future commits:

**Open questions / TODOs**
- [x] ~~How should serialization/deserialization work ideally? Should the storage be free to choose the format or should we provide a serialized form, possibly customizable so that the storage can decide which fields to persist?~~
- [x] ~~For deserialization of `EngineSession` we definitely need enhancements to the engine modules as only the engine knows how to contruct and instance of an `EngineSession`~~
- [ ] Should we introduce a `concept-session-storage` and `feature-session-storage` module to contain the interface contract? As it stands, a custom storage module would have to depend on `feature-session`, which seems fine, but more granularity might be better?
- [ ] Let's create a new type `SessionPair` instead of using `SessionProxy` for storage. `SessionProxy` can then also use `SessionPair`
- [ ] There's now duplicated state between SessionProvider/Storage and manager which needs to be refactored.

csadilek commented:
Sputnik is not responding as it tries to run analysis on the removed/renamed files, which seems wrong. I filed a bug: https://github.com/TouK/sputnik-ci/issues/22

pocmo commented on components/feature/session/src/main/java/mozilla/components/feature/session/DefaultSessionStorage.kt:
```diff
@@ -0,0 +1,139 @@
+ * }
+ */
+class DefaultSessionStorage(private val context: Context) : SessionStorage {
+    private val defaultFileName = DefaultSessionStorage::class.java.canonicalName.replace(".", "_") + ".json"
```
Do we really want to migrate this file whenever we rename or move this class? :)

pocmo commented on components/feature/session/src/main/java/mozilla/components/feature/session/DefaultSessionStorage.kt:
```diff
@@ -0,0 +1,139 @@
+
+    @Synchronized
+    override fun add(session: SessionProxy): String {
+        val id = UUID.randomUUID().toString()
```
I think every session should contain a UUID. We do this in Focus and it's a pretty handy identifier to pass around over app boundaries (e.g. in notifications) when you need to reference a specific tab/session. That's something we can do in a follow-up though. :)

pocmo commented on components/feature/session/src/main/java/mozilla/components/feature/session/DefaultSessionStorage.kt:
```diff
@@ -0,0 +1,139 @@
+class DefaultSessionStorage(private val context: Context) : SessionStorage {
+    private val defaultFileName = DefaultSessionStorage::class.java.canonicalName.replace(".", "_") + ".json"
+    private val sessions = mutableMapOf<String, SessionProxy>()
+    private var selectedSession: Session? = null
```
When refactoring this to not have the mapping here then we should make sure to avoid mirroring the selection here too. (Let's file issues for all the refactorings we want to do).

csadilek replied:
OK, updated to use const file name.

pocmo commented on components/feature/session/src/main/java/mozilla/components/feature/session/DefaultSessionStorage.kt:
```diff
@@ -0,0 +1,139 @@
+ *         "session": {}
+ *         "engineSession": {}
+ *     }
+ * }
```
Let's also add some kind of version int somewhere in here. At some future point we might need to do some migrations and/or clean ups.

csadilek replied:
good catch, done.

pocmo commented on components/feature/session/src/main/java/mozilla/components/feature/session/SessionStorage.kt:
```diff
@@ -0,0 +1,64 @@
+     * sessions.
+     * @return list of all restored sessions
+     */
+    fun restore(engine: Engine): List<SessionProxy>
```
Just a thought: I was wondering whether this should just restore some "Snapshot" object that contains everything (like the serialized data) instead of having multiple getters.

csadilek commented:
@pocmo OK, addressed some comments and filed follow-up issues. Mostly I think introducing the SessionPair (#163) will clean things up quite a bit. I will work on that next.

csadilek commented:
This one needs to be merged manually though, due to the Sputnik problem.
